packages/mattermost-redux/src/types/actions.ts

```diff
@@ -12,10 +12,22 @@ export type Thunk = (b: DispatchFunc, a: GetStateFunc) => Promise<ActionResult>
 
 export type Action = GenericAction | Thunk | BatchAction | ActionFunc;
 
-export type ActionResult = {
-    data?: any;
-    error?: any;
+export type ActionResult<Data = any, Error = any> = {
+    data?: Data;
+    error?: Error;
 };
 
 export type DispatchFunc = (action: Action, getState?: GetStateFunc | null) => Promise<ActionResult>;
-export type ActionFunc = (dispatch: DispatchFunc, getState: GetStateFunc) => Promise<ActionResult|ActionResult[]> | ActionResult;
+
+/**
+ * Return type of a redux action.
+ * @usage
+ * ActionFunc<ReturnTypeOfData, ErrorTypeReturnedFromServer>
+ */
+export type ActionFunc<Data = any, Error = any> = (
```

Comment:
Nothing to change, but I was thinking about possibly including the type of the state here so that we could also parameterize `getState`. I was debating on if it makes sense to include a `State` type before `Data` and `Error` since they come earlier in the type signature, but the more I think about it, the more I expect `Data` to be the most used type argument

Reply:
Also GetStateFunc return GlobalState of @mattermost/types/store shouldn't it be of "types/store" ? Which could be hard since these types are defined in mm-redux package.

Reply:
I think what we want is to make `GetStateFunc` parameterized so that it can return either version of `GlobalState` depending on where the code is located. The default case would be for it to use the mattermost-redux version and we could override it to return the web app's version. That would mean that we need a third type argument here, but that's what we'll need eventually if we ever want this to converge with the proper `ThunkAction` type from react-redux ([link](https://github.com/reduxjs/redux-thunk/blob/master/src/types.ts#L52)). That's something I've wanted to do for a while, but I haven't committed to doing it yet because I still have some doubts about how to type the whole state tree.

For example, I feel like it might be better to have a bunch of smaller types that represent the different parts of the state tree that might be lazy loaded like `state.views` for the web app, `state.boards`, etc for different parts of the app. At the same time, it seems like it'd be better to just write `type GlobalState = ReturnType<typeof store.getState>`, but that also will probably reveal a bunch of issues with how we've typed various things as well. Anyway, this second paragraph is just rambling

Reply:
Opened ticket to ponder over the issue more later https://mattermost.atlassian.net/browse/MM-46908